File: NetFramework/Pdf/CS/CoolPrintPreview/CoolPrintPreviewDialog.cs
﻿using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Drawing.Printing;
using System.Text;
using System.Windows.Forms;

namespace CoolPrintPreview
{
    /// <summary>
    /// Represents a dialog containing a <see cref="CoolPrintPreviewControl"/> control
    /// used to preview and print <see cref="PrintDocument"/> objects.
    /// </summary>
    /// <remarks>
    /// This dialog is similar to the standard <see cref="PrintPreviewDialog"/>
    /// but provides additional options such printer and page setup buttons,
    /// a better UI based on the <see cref="ToolStrip"/> control, and built-in
    /// PDF export.
    /// </remarks>
    internal partial class CoolPrintPreviewDialog : Form
    {
        //--------------------------------------------------------------------
        #region ** fields

        PrintDocument _doc;

        #endregion

        //--------------------------------------------------------------------
        #region ** ctor

        /// <summary>
        /// Initializes a new instance of a <see cref="CoolPrintPreviewDialog"/>.
        /// </summary>
        public CoolPrintPreviewDialog() : this(null)
        {
        }
        /// <summary>
        /// Initializes a new instance of a <see cref="CoolPrintPreviewDialog"/>.
        /// </summary>
        /// <param name="parentForm">Parent form that defines the initial size for this dialog.</param>
        public CoolPrintPreviewDialog(Control parentForm)
        {
            InitializeComponent();
            if (parentForm != null)
            {
                Size = parentForm.Size;
            }
        }
        #endregion

        //--------------------------------------------------------------------
        #region ** object model

        /// <summary>
        /// Gets or sets the <see cref="PrintDocument"/> to preview.
        /// </summary>
        public PrintDocument Document
        {
            get { return _doc; }
            set
            {
                // unhook event handlers
                if (_doc != null)
                {
                    _doc.BeginPrint -= _doc_BeginPrint;
                    _doc.EndPrint -= _doc_EndPrint;
                }

                // save the value
                _doc = value;

                // hook up event handlers
                if (_doc != null)
                {
                    _doc.BeginPrint += _doc_BeginPrint;
                    _doc.EndPrint += _doc_EndPrint;
                }


                // don't assign document to preview until this form becomes visible
                if (Visible)
                {
                    _preview.Document = Document;
                }
            }
        }

        #endregion

        //--------------------------------------------------------------------
        #region ** overloads

        /// <summary>
        /// Overridden to assign document to preview control only after the 
        /// initial activation.
        /// </summary>
        /// <param name="e"><see cref="EventArgs"/> that contains the event data.</param>
        protected override void OnShown(EventArgs e)
        {
            base.OnShown(e);
            _preview.Document = Document;
        }
        /// <summary>
        /// Overridden to cancel any ongoing previews when closing form.
        /// </summary>
        /// <param name="e"><see cref="FormClosingEventArgs"/> that contains the event data.</param>
        protected override void OnFormClosing(FormClosingEventArgs e)
        {
            base.OnFormClosing(e);
            if (_preview.IsRendering && !e.Cancel)
            {
                _preview.Cancel();
            }
        }

        #endregion

        //--------------------------------------------------------------------
        #region ** main commands

        void _btnCreatePdf_Click(object sender, EventArgs e)
        {
            using (var dlg = new SaveFileDialog())
            {
                dlg.Filter = "Portable Document File (*.pdf)|*.pdf";
                dlg.DefaultExt = ".pdf";
                if (dlg.ShowDialog() == DialogResult.OK)
                {
                    var pdfGenerator = new PrintDocumentPdfExporter(_pdf);
                    if (pdfGenerator.RenderDocument(Document, true))
                    {
                        _pdf.Save(dlg.FileName);
                    }
                }
            }
        }
        void _btnPrint_Click(object sender, EventArgs e)
        {
            using (var dlg = new PrintDialog())
            {
                // configure dialog
                dlg.AllowSomePages = true;
                dlg.AllowSelection = true;
                dlg.UseEXDialog = true;
                dlg.Document = Document;

                // show allowed page range
                var ps = dlg.PrinterSettings;
                ps.MinimumPage = ps.FromPage = 1;
                ps.MaximumPage = ps.ToPage = _preview.PageCount;

                // show dialog
                if (dlg.ShowDialog(this) == DialogResult.OK)
                {
                    // print selected page range
                    _preview.Print();
                }
            }
        }
        void _btnPageSetup_Click(object sender, EventArgs e)
        {
            using (var dlg = new PageSetupDialog())
            {
                dlg.Document = Document;
                if (dlg.ShowDialog(this) == DialogResult.OK)
                {
                    // to show new page layout
                    _preview.RefreshPreview();
                }
            }
        }

        #endregion

        //--------------------------------------------------------------------
        #region ** zoom

        void _btnZoom_ButtonClick(object sender, EventArgs e)
        {
            _preview.ZoomMode = _preview.ZoomMode == ZoomMode.ActualSize
                ? ZoomMode.FullPage
                : ZoomMode.ActualSize;
        }
        void _btnZoom_DropDownItemClicked(object sender, ToolStripItemClickedEventArgs e)
        {
            if (e.ClickedItem == _itemActualSize)
            {
                _preview.ZoomMode = ZoomMode.ActualSize;
            }
            else if (e.ClickedItem == _itemFullPage)
            {
                _preview.ZoomMode = ZoomMode.FullPage;
            }
            else if (e.ClickedItem == _itemPageWidth)
            {
                _preview.ZoomMode = ZoomMode.PageWidth;
            }
            else if (e.ClickedItem == _itemTwoPages)
            {
                _preview.ZoomMode = ZoomMode.TwoPages;
            }
            if (e.ClickedItem == _item10)
            {
                _preview.Zoom = .1;
            }
            else if (e.ClickedItem == _item100)
            {
                _preview.Zoom = 1;
            }
            else if (e.ClickedItem == _item150)
            {
                _preview.Zoom = 1.5;
            }
            else if (e.ClickedItem == _item200)
            {
                _preview.Zoom = 2;
            }
            else if (e.ClickedItem == _item25)
            {
                _preview.Zoom = .25;
            }
            else if (e.ClickedItem == _item50)
            {
                _preview.Zoom = .5;
            }
            else if (e.ClickedItem == _item500)
            {
                _preview.Zoom = 5;
            }
            else if (e.ClickedItem == _item75)
            {
                _preview.Zoom = .75;
            }
        }
        #endregion

        //--------------------------------------------------------------------
        #region ** page navigation

        void _btnFirst_Click(object sender, EventArgs e)
        {
            _preview.StartPage = 0;
        }
        void _btnPrev_Click(object sender, EventArgs e)
        {
            _preview.StartPage--;
        }
        void _btnNext_Click(object sender, EventArgs e)
        {
            _preview.StartPage++;
        }
        void _btnLast_Click(object sender, EventArgs e)
        {
            _preview.StartPage = _preview.PageCount - 1;
        }
        void _txtStartPage_Enter(object sender, EventArgs e)
        {
            _txtStartPage.SelectAll();
        }
        void _txtStartPage_Validating(object sender, CancelEventArgs e)
        {
            CommitPageNumber();
        }
        void _txtStartPage_KeyPress(object sender, KeyPressEventArgs e)
        {
            var c = e.KeyChar;
            if (c == (char)13)
            {
                CommitPageNumber();
                e.Handled = true;
            }
            else if (c > ' ' && !char.IsDigit(c))
            {
                e.Handled = true;
            }
        }
        void CommitPageNumber()
        {
            int page;
            if (int.TryParse(_txtStartPage.Text, out page))
            {
                _preview.StartPage = page - 1;
            }
        }
        void _preview_StartPageChanged(object sender, EventArgs e)
        {
            var page = _preview.StartPage + 1;
            _txtStartPage.Text = page.ToString();
        }
        private void _preview_PageCountChanged(object sender, EventArgs e)
        {
            this.Update();
            Application.DoEvents();
            _lblPageCount.Text = string.Format("of {0}", _preview.PageCount);
        }

        #endregion

        //--------------------------------------------------------------------
        #region ** job control

        void _btnCancel_Click(object sender, EventArgs e)
        {
            if (_preview.IsRendering)
            {
                _preview.Cancel();
            }
            else
            {
                Close();
            }
        }
        void _doc_BeginPrint(object sender, PrintEventArgs e)
        {
            _btnCancel.Text = "&Cancel";
            _btnPrint.Enabled = _btnCreatePdf.Enabled = _btnPageSetup.Enabled = false;
        }
        void _doc_EndPrint(object sender, PrintEventArgs e)
        {
            _btnCancel.Text = "&Close";
            _btnPrint.Enabled = _btnCreatePdf.Enabled = _btnPageSetup.Enabled = true;
        }

        #endregion
    }
}

File: NetFramework/Pdf/VB/CoolPrintPreview/CoolPrintPreviewDialog.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Drawing.Printing
Imports System.Text
Imports System.Windows.Forms

''' <summary>
''' Represents a dialog containing a <see cref="CoolPrintPreviewControl"/> control
''' used to preview and print <see cref="PrintDocument"/> objects.
''' </summary>
''' <remarks>
''' This dialog is similar to the standard <see cref="PrintPreviewDialog"/>
''' but provides additional options such printer and page setup buttons,
''' a better UI based on the <see cref="ToolStrip"/> control, and built-in
''' PDF export.
''' </remarks>
Friend Partial Class CoolPrintPreviewDialog
	Inherits Form
	'--------------------------------------------------------------------
	#Region "** fields"

	Private _doc As PrintDocument

	#End Region

	'--------------------------------------------------------------------
	#Region "** ctor"

	''' <summary>
	''' Initializes a new instance of a <see cref="CoolPrintPreviewDialog"/>.
	''' </summary>
	Public Sub New()
		Me.New(Nothing)
	End Sub
	''' <summary>
	''' Initializes a new instance of a <see cref="CoolPrintPreviewDialog"/>.
	''' </summary>
	''' <param name="parentForm">Parent form that defines the initial size for this dialog.</param>
	Public Sub New(parentForm As Control)
		InitializeComponent()
		If parentForm IsNot Nothing Then
			Size = parentForm.Size
		End If
	End Sub
	#End Region

	'--------------------------------------------------------------------
	#Region "** object model"

	''' <summary>
	''' Gets or sets the <see cref="PrintDocument"/> to preview.
	''' </summary>
	Public Property Document() As PrintDocument
		Get
			Return _doc
		End Get
		Set
			' unhook event handlers
			If _doc IsNot Nothing Then
				RemoveHandler _doc.BeginPrint, AddressOf _doc_BeginPrint
				RemoveHandler _doc.EndPrint, AddressOf _doc_EndPrint
			End If

			' save the value
			_doc = value

			' hook up event handlers
			If _doc IsNot Nothing Then
				AddHandler _doc.BeginPrint, AddressOf _doc_BeginPrint
				AddHandler _doc.EndPrint, AddressOf _doc_EndPrint
			End If


			' don't assign document to preview until this form becomes visible
			If Visible Then
				_preview.Document = Document
			End If
		End Set
	End Property

	#End Region

	'--------------------------------------------------------------------
	#Region "** overloads"

	''' <summary>
	''' Overridden to assign document to preview control only after the 
	''' initial activation.
	''' </summary>
	''' <param name="e"><see cref="EventArgs"/> that contains the event data.</param>
	Protected Overrides Sub OnShown(e As EventArgs)
		MyBase.OnShown(e)
		_preview.Document = Document
	End Sub
	''' <summary>
	''' Overridden to cancel any ongoing previews when closing form.
	''' </summary>
	''' <param name="e"><see cref="FormClosingEventArgs"/> that contains the event data.</param>
	Protected Overrides Sub OnFormClosing(e As FormClosingEventArgs)
		MyBase.OnFormClosing(e)
		If _preview.IsRendering AndAlso Not e.Cancel Then
			_preview.Cancel()
		End If
	End Sub

	#End Region

	'--------------------------------------------------------------------
	#Region "** main commands"

	Private Sub _btnCreatePdf_Click(sender As Object, e As EventArgs)
        Dim dlg As New SaveFileDialog()
        dlg.Filter = "Portable Document File (*.pdf)|*.pdf"
        dlg.DefaultExt = ".pdf"
        If dlg.ShowDialog() = DialogResult.OK Then
            Dim pdfGenerator As New PrintDocumentPdfExporter(_pdf)
            If pdfGenerator.RenderDocument(Document, True) Then
                _pdf.Save(dlg.FileName)
            End If
        End If
    End Sub
	Private Sub _btnPrint_Click(sender As Object, e As EventArgs)
        Dim dlg As New PrintDialog()

        ' configure dialog
        dlg.AllowSomePages = True
        dlg.AllowSelection = True
        dlg.UseEXDialog = True
        dlg.Document = Document

        ' show allowed page range
        Dim ps As PrinterSettings = dlg.PrinterSettings
        ps.MinimumPage = InlineAssignHelper(ps.FromPage, 1)
        ps.MaximumPage = InlineAssignHelper(ps.ToPage, _preview.PageCount)

        ' show dialog
        If dlg.ShowDialog(Me) = DialogResult.OK Then
            ' print selected page range
            _preview.Print()
        End If
    End Sub
	Private Sub _btnPageSetup_Click(sender As Object, e As EventArgs)
        Dim dlg As New PageSetupDialog()
        dlg.Document = Document
        If dlg.ShowDialog(Me) = DialogResult.OK Then
            ' to show new page layout
            _preview.RefreshPreview()
        End If
    End Sub

	#End Region

	'--------------------------------------------------------------------
	#Region "** zoom"

	Private Sub _btnZoom_ButtonClick(sender As Object, e As EventArgs)
		_preview.ZoomMode = If(_preview.ZoomMode = ZoomMode.ActualSize, ZoomMode.FullPage, ZoomMode.ActualSize)
	End Sub
	Private Sub _btnZoom_DropDownItemClicked(sender As Object, e As ToolStripItemClickedEventArgs)
		If e.ClickedItem Is _itemActualSize Then
			_preview.ZoomMode = ZoomMode.ActualSize
		ElseIf e.ClickedItem Is _itemFullPage Then
			_preview.ZoomMode = ZoomMode.FullPage
		ElseIf e.ClickedItem Is _itemPageWidth Then
			_preview.ZoomMode = ZoomMode.PageWidth
		ElseIf e.ClickedItem Is _itemTwoPages Then
			_preview.ZoomMode = ZoomMode.TwoPages
		End If
		If e.ClickedItem Is _item10 Then
			_preview.Zoom = 0.1
		ElseIf e.ClickedItem Is _item100 Then
			_preview.Zoom = 1
		ElseIf e.ClickedItem Is _item150 Then
			_preview.Zoom = 1.5
		ElseIf e.ClickedItem Is _item200 Then
			_preview.Zoom = 2
		ElseIf e.ClickedItem Is _item25 Then
			_preview.Zoom = 0.25
		ElseIf e.ClickedItem Is _item50 Then
			_preview.Zoom = 0.5
		ElseIf e.ClickedItem Is _item500 Then
			_preview.Zoom = 5
		ElseIf e.ClickedItem Is _item75 Then
			_preview.Zoom = 0.75
		End If
	End Sub
	#End Region

	'--------------------------------------------------------------------
	#Region "** page navigation"

	Private Sub _btnFirst_Click(sender As Object, e As EventArgs)
		_preview.StartPage = 0
	End Sub
	Private Sub _btnPrev_Click(sender As Object, e As EventArgs)
		_preview.StartPage -= 1
	End Sub
	Private Sub _btnNext_Click(sender As Object, e As EventArgs)
		_preview.StartPage += 1
	End Sub
	Private Sub _btnLast_Click(sender As Object, e As EventArgs)
		_preview.StartPage = _preview.PageCount - 1
	End Sub
	Private Sub _txtStartPage_Enter(sender As Object, e As EventArgs)
		_txtStartPage.SelectAll()
	End Sub
	Private Sub _txtStartPage_Validating(sender As Object, e As CancelEventArgs)
		CommitPageNumber()
	End Sub
	Private Sub _txtStartPage_KeyPress(sender As Object, e As KeyPressEventArgs)
        Dim c As Char = e.KeyChar
		If c = ChrW(13) Then
			CommitPageNumber()
			e.Handled = True
		ElseIf c > " "C AndAlso Not Char.IsDigit(c) Then
			e.Handled = True
		End If
	End Sub
	Private Sub CommitPageNumber()
		Dim page As Integer
		If Integer.TryParse(_txtStartPage.Text, page) Then
			_preview.StartPage = page - 1
		End If
	End Sub
	Private Sub _preview_StartPageChanged(sender As Object, e As EventArgs)
        Dim page As Integer = _preview.StartPage + 1
		_txtStartPage.Text = page.ToString()
	End Sub
	Private Sub _preview_PageCountChanged(sender As Object, e As EventArgs)
		Me.Update()
		Application.DoEvents()
		_lblPageCount.Text = String.Format("of {0}", _preview.PageCount)
	End Sub

	#End Region

	'--------------------------------------------------------------------
	#Region "** job control"

	Private Sub _btnCancel_Click(sender As Object, e As EventArgs)
		If _preview.IsRendering Then
			_preview.Cancel()
		Else
			Close()
		End If
	End Sub
	Private Sub _doc_BeginPrint(sender As Object, e As PrintEventArgs)
		_btnCancel.Text = "&Cancel"
		_btnPrint.Enabled = InlineAssignHelper(_btnCreatePdf.Enabled, InlineAssignHelper(_btnPageSetup.Enabled, False))
	End Sub
	Private Sub _doc_EndPrint(sender As Object, e As PrintEventArgs)
		_btnCancel.Text = "&Close"
		_btnPrint.Enabled = InlineAssignHelper(_btnCreatePdf.Enabled, InlineAssignHelper(_btnPageSetup.Enabled, True))
	End Sub
	Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
		target = value
		Return value
	End Function

	#End Region
End Class
